models/user.js

```diff
@@ -18,11 +20,46 @@ const userSchema = new mongoose.Schema({
     required: true,
     validate: {
       validator(v) {
-        return /^(ftp|http|https):\/\/[^ "]+$/.test(v);
+        // return /^(ftp|http|https):\/\/[^ "]+$/.test(v);
+        return isURL(v);
       },
       message: (props) => `${props.value} is not a valid link!`,
     },
   },
+  email: {
+    type: String,
+    required: true,
+    unique: true,
+    validate: {
+      validator(v) {
+        return isEmail(v);
+      },
+      message: (props) => `${props.value} is not a valid email!`,
+    },
+  },
+  password: {
+    type: String,
+    required: true,
+    minlength: 8,
+    select: false,
+  },
 });
 
+// eslint-disable-next-line func-names
+userSchema.statics.findUserByCredentials = function (email, password) {
+  return this.findOne({ email }).select('+password')
+    .then((user) => {
+      if (!user) {
+        return Promise.reject(new Error('Неправильная почта или пароль!'));
```

Comment:
Можно лучше: перенести класс ошибки из контроллера.
https://github.com/suholet/mesto-server/blob/f6b256c0eb9092965422a8b25ee3863034ddfd90/controllers/users.js#L119-L122
В этом случае можно `.catch()` блок можно заменить на `.catch(next);`, а ошибку с кодом 401 бросать прямо здесь:
```suggestion
        return Promise.reject(new UnathorizedError('Неправильная почта или пароль!'));
```